### PR TITLE
Feature/new server infra

### PR DESCRIPTION
**Pull request by lamai93**

### CODE REVIEW

**Comment by lamai93** [comment]
http://jenkins01.arangodb.biz:8080/job/arangodb-matrix-pr/2499/

**Comment by lamai93** [comment]
http://jenkins01.arangodb.biz:8080/job/arangodb-matrix-pr/2532/

File: arangod/Scheduler/SupervisedScheduler.cpp [hunk 1]
```diff
@@ -104,11 +101,17 @@ SupervisedScheduler::SupervisedScheduler(uint64_t minThreads,
 
 SupervisedScheduler::~SupervisedScheduler() {}
 
+bool SupervisedScheduler::queue(RequestLane lane, std::function<void()> const& handler) {
+  std::function<void()> copy = handler;
+  return queue(lane, std::move(copy));
```
Comment (mpoeter) [inline arangod/Scheduler/SupervisedScheduler.cpp]: You could change the signature of `queue` to take `handler` per value, then you would't need a separate overload - r-values would be moved, l-values (like const refs) would be copied.

File: arangod/Pregel/Conductor.cpp [hunk 2]
```diff
@@ -772,8 +769,8 @@ int Conductor::_sendToAllDBServers(std::string const& path,
       handle(response.slice());
     } else {
       TRI_ASSERT(SchedulerFeature::SCHEDULER != nullptr);
-      rest::Scheduler* scheduler = SchedulerFeature::SCHEDULER;
-      scheduler->queue(RequestPriority::LOW, [this, path, message] {
+      Scheduler* scheduler = SchedulerFeature::SCHEDULER;
+      scheduler->queue(RequestLane::CLUSTER_INTERNAL, [this, path, message] {
```
Comment (neunhoef) [inline arangod/Pregel/Conductor.cpp]: Should this be INTERNAL_LOW?

File: arangod/RestHandler/RestShutdownHandler.cpp [hunk 3]
```diff
@@ -92,23 +92,23 @@ RestStatus RestShutdownHandler::execute() {
...
     generateResult(rest::ResponseCode::OK, result.slice());
   } catch (...) {
     // Ignore the error
   }
-  rest::Scheduler* scheduler = SchedulerFeature::SCHEDULER;
+  Scheduler* scheduler = SchedulerFeature::SCHEDULER;
   // don't block the response for workers waiting on this callback
   // this should allow workers to go into the IDLE state
-  scheduler->queue(RequestPriority::HIGH, [this] {
+  scheduler->queue(RequestLane::CLUSTER_INTERNAL, [this] {
```
Comment (neunhoef) [inline arangod/RestHandler/RestShutdownHandler.cpp]: This sounds all right, cluster internal is higher priority.

File: arangod/VocBase/Methods/Tasks.cpp [hunk 4]
```diff
@@ -258,7 +258,7 @@ std::function<void(const asio::error_code&)> Task::callbackFunction() {
         // remove task from list of tasks if it is still active
         if (this == (*itr).second.second.get()) {
           // still the same task. must remove from map
-          _tasks.erase(itr);
+          //_tasks.erase(itr); --- this invalidates the iterator in Task::shutdownTasks
```
Comment (neunhoef) [inline arangod/VocBase/Methods/Tasks.cpp]: Where is this cleanup now?

File: arangod/VocBase/Methods/Tasks.cpp [hunk 5]
```diff
@@ -276,26 +276,25 @@ std::function<void(const asio::error_code&)> Task::callbackFunction() {
...
 
-          if (_periodic.load() && !SchedulerFeature::SCHEDULER->isStopping()) {
+          if (_periodic.load() && !application_features::ApplicationServer::isStopping()) {
             // requeue the task
             queue(_interval);
           } else {
             // in case of one-off tasks or in case of a shutdown, simply
             // remove the task from the list
-            Task::unregisterTask(_id, false);
+            //Task::unregisterTask(_id, false);
```
Comment (neunhoef) [inline arangod/VocBase/Methods/Tasks.cpp]: Where is this now?

File: etc/testing/arangod.conf [hunk 6]
```diff
@@ -1 +1,3 @@
 @include etc/testing/arangod-common.conf
+[log]
+level=debug
```
Comment (neunhoef) [inline etc/testing/arangod.conf]: This is too much in general. Please remove again.

File: arangod/Pregel/Conductor.cpp (hunk 2 above)
Comment (lamai93) [reply arangod/Pregel/Conductor.cpp]: Yes

File: arangod/GeneralServer/ListenTask.cpp [hunk 7]
```diff
@@ -78,10 +77,6 @@ void ListenTask::accept() {
 
   auto handler = [this, self](asio_ns::error_code const& ec) {
 
-    if (!_bound) {
-      _handler = nullptr;
-      return;
-    }
```
[the remark this answers is not in the document]
Comment (lamai93) [reply arangod/GeneralServer/ListenTask.cpp]: `_handler` was a member of ListenTask. Now it is a local variable that is passed to `asyncAccept`. In `ListenTask::stop` the `asyncAccept` is aborted. In this case the handler does not call `accept` again. Hence it is no longer needed.

File: arangod/Scheduler/Scheduler.cpp [hunk 8]
```diff
@@ -32,674 +32,149 @@
...
-  uint64_t const nrRunning = numRunning(counters);
+  TRI_ASSERT(!isStopping());
 
-  if (nrRunning <= _minThreads) {
-    // don't stop a thread if we already reached the minimum
-    // number of threads
-    return false;
+  if (delay < std::chrono::milliseconds(1)) {
+    // execute directly
+    queue(lane, [handler](){ handler(false); });
```
Comment (mpoeter) [inline arangod/Scheduler/Scheduler.cpp]: A copy of the `handler` is queued directly here and then another in the `_cronQueue` a few lines down. Is this intentional? This causes the handler to be run twice.

File: arangod/Scheduler/SchedulerFeature.cpp [hunk 9]
```diff
@@ -47,24 +49,41 @@ using namespace arangodb::basics;
...
 using namespace arangodb::rest;
 
+namespace {
+  /// @brief return the default number of threads to use (upper bound)
+  size_t defaultNumberOfThreads() {
+    // use two times the number of hardware threads as the default
+    size_t result = TRI_numberProcessors() * 2;
+    // but only if higher than 64. otherwise use a default minimum value of 64
+    if (result < 64) {
+      result = 64;
```
Comment (mpoeter) [inline arangod/Scheduler/SchedulerFeature.cpp]: IMHO 64 is way to high for a conventional machine (e.g. 4 cores).

File: arangod/Scheduler/SupervisedScheduler.cpp [hunk 10]
```diff
@@ -0,0 +1,490 @@
...
+
+  {
+    // inform the supervisor that this thread is alive
+    std::lock_guard<std::mutex> guard(_mutexSupervisor);
+    id = _numWorker++;  // increase the number of workers here, to obtains the id
+    _conditionSupervisor.notify_one();
+  }
+
+  // copy shared_ptr
+  auto state = _workerStates.back();
```
Comment (mpoeter) [inline arangod/Scheduler/SupervisedScheduler.cpp]: This looks like an unsynchronized access to `_workerStates`.

File: arangod/Scheduler/SchedulerFeature.cpp (hunk 9 above)
Comment (lamai93) [reply arangod/Scheduler/SchedulerFeature.cpp]: Not with out blocking threads arch in arangodb. This is required until things like ClusterComm are completely async.

Comment (mpoeter) [reply arangod/Scheduler/SchedulerFeature.cpp]: I know that threads can block in the current architecture, and you need additional threads to compensate this. You obviously have more insight than me, but 64 still feels pretty high. I would probably cap the number of threads by 8 per core.

**Comment by lamai93** [comment]
http://jenkins01.arangodb.biz:8080/job/arangodb-matrix-pr/2670/

File: arangod/Pregel/Conductor.cpp (hunk 2 above)
Comment (neunhoef) [reply arangod/Pregel/Conductor.cpp]: And?

File: arangod/Scheduler/SchedulerFeature.cpp [hunk 11]
```diff
@@ -332,61 +403,10 @@ void SchedulerFeature::buildControlCHandler() {
...
-
-    LOG_TOPIC(INFO, arangodb::Logger::FIXME)
-        << "hangup received, about to reopen logfile";
-    LogAppender::reopen();
-    LOG_TOPIC(INFO, arangodb::Logger::FIXME)
-        << "hangup received, reopened logfile";
-
-    _hangupSignals->async_wait(_hangupHandler);
-  };
+  std::signal(SIGINT, c_exit_handler);
```
Comment (neunhoef) [inline arangod/Scheduler/SchedulerFeature.cpp]: I think we should use `sigaction` instead.

**Comment by lamai93** [comment]
http://jenkins01.arangodb.biz:8080/job/arangodb-matrix-pr/2673/

**Pull request merged**